compiler/src/iree/compiler/Dialect/Stream/Transforms/test/unify_encoding_for_globals.mlir

```diff
@@ -38,14 +42,214 @@ module @immutable_source_with_initial_value {
+
+  util.return
+}
+
+// -----
+
+// Test: multiple devices with different resolvers encoding the same source global.
+// Since different resolvers produce different encodings and they share the same source,
+// there's no common encoding - should fall back to identity encoding.
+
+#executable_target_0 = #hal.executable.target<"vmvx", "vmvx-bytecode-fb", {iree.encoding.resolver = #iree_encoding.specialization_resolver<123>}>
+#executable_target_1 = #hal.executable.target<"vmvx", "vmvx-bytecode-fb", {iree.encoding.resolver = #iree_encoding.specialization_resolver<456>}>
+#device_target_local_0 = #hal.device.target<"local", {ordinal = 0 : index}, [#executable_target_0]> : !hal.device
+#device_target_local_1 = #hal.device.target<"local", {ordinal = 1 : index}, [#executable_target_1]> : !hal.device
+#encoding1 = #iree_encoding.testing<layouts = [#iree_encoding.specialized<111>]>
+#encoding2 = #iree_encoding.testing<layouts = [#iree_encoding.specialized<222>]>
+
+// CHECK: util.global private @[[$DEVICE_A:.+]] =
+// CHECK: util.global private @[[$DEVICE_B:.+]] =
+util.global private @device_a = #device_target_local_0
+util.global private @device_b = #device_target_local_1
+// Single source global shared by both devices.
+util.global private @weight : !stream.resource<constant>
+util.global private @weight_size : index
+util.global private @encoded_a_v1 : !stream.resource<constant>
+util.global private @encoded_a_v2 : !stream.resource<constant>
+util.global private @encoded_b_v1 : !stream.resource<constant>
+util.global private @encoded_b_v2 : !stream.resource<constant>
+
+// CHECK: util.initializer
+util.initializer {
+  %cst = stream.tensor.constant on(#hal.device.affinity<@device_a>) : tensor<4096x4096xf32> in !stream.resource<constant> = #stream.parameter.named<"model"::"weight"> : tensor<4096x4096xf32>
+  %0 = stream.resource.size %cst : !stream.resource<constant>
+  util.global.store %cst, @weight : !stream.resource<constant>
+  util.global.store %0, @weight_size : index
+
+  // CHECK: %[[SOURCE:.+]] = util.global.load @weight
+  %source = util.global.load @weight : !stream.resource<constant>
+  %source_size = util.global.load @weight_size : index
+
+  // Device A encodes the shared source - should get identity encoding.
+  // CHECK: stream.tensor.sizeof on(#hal.device.affinity<@[[$DEVICE_A]]>) tensor<4096x4096xf32, #iree_encoding.identity>
+  // CHECK: stream.tensor.encode on(#hal.device.affinity<@[[$DEVICE_A]]>) %[[SOURCE]] : {{.*}} -> tensor<4096x4096xf32, #iree_encoding.identity>
+  %size1 = stream.tensor.sizeof on(#hal.device.affinity<@device_a>) tensor<4096x4096xf32, #encoding1> : index
+  %enc1 = stream.tensor.encode on(#hal.device.affinity<@device_a>) %source : tensor<4096x4096xf32> in !stream.resource<constant>{%source_size} -> tensor<4096x4096xf32, #encoding1> in !stream.resource<*>{%size1}
+  %const1 = stream.async.clone on(#hal.device.affinity<@device_a>) %enc1 : !stream.resource<*>{%size1} -> !stream.resource<constant>{%size1}
```

Comment:
Do we need all the `clone`s here? It increases the test code size and correct handling of `clone` is/should be tested by other tests.

Reply:
Thanks for pointing it out! I removed the `clone`s.